rustls/src/crypto/cipher.rs

```diff
@@ -3,11 +3,69 @@ use crate::error::{Error, PeerMisbehaved};
 use crate::msgs::codec;
 use crate::msgs::fragmenter::MAX_FRAGMENT_LEN;
 pub use crate::msgs::message::{BorrowedPlainMessage, OpaqueMessage, PlainMessage};
+#[cfg(feature = "secret_extraction")]
+use crate::suites::ConnectionTrafficSecrets;
+
+/// Factory trait for building `MessageEncrypter` and `MessageDecrypter` for a TLS1.3 cipher suite.
+pub trait Tls13AeadAlgorithm: Send + Sync {
+    /// Build a `MessageEncrypter` for the given key/iv.
+    fn encrypter(&self, key: AeadKey, iv: Iv) -> Box<dyn MessageEncrypter>;
+
+    /// Build a `MessageDecrypter` for the given key/iv.
+    fn decrypter(&self, key: AeadKey, iv: Iv) -> Box<dyn MessageDecrypter>;
+
+    /// The length of key in bytes required by `encrypter()` and `decrypter()`.
+    fn key_len(&self) -> usize;
+
+    #[cfg(feature = "secret_extraction")]
+    /// Convert the key material from `key`/`iv`, into a `ConnectionTrafficSecrets` item.
+    fn extract_keys(&self, key: AeadKey, iv: Iv) -> ConnectionTrafficSecrets;
+}
+
+/// Factory trait for building `MessageEncrypter` and `MessageDecrypter` for a TLS1.2 cipher suite.
+pub trait Tls12AeadAlgorithm: Send + Sync + 'static {
+    /// Build a `MessageEncrypter` for the given key/iv and extra key block (which can be used for
+    /// improving explicit nonce size security, if needed).
+    fn encrypter(&self, key: &[u8], iv: &[u8], extra: &[u8]) -> Box<dyn MessageEncrypter>;
+
+    /// Build a `MessageDecrypter` for the given key/iv.
+    fn decrypter(&self, key: &[u8], iv: &[u8]) -> Box<dyn MessageDecrypter>;
+
+    /// Return a `KeyBlockShape` that defines how large the `key_block` is and how it
+    /// is split up prior to calling `encrypter()`, `decrypter()` and/or `extract_keys()`.
+    fn key_block_shape(&self) -> KeyBlockShape;
+
+    #[cfg(feature = "secret_extraction")]
+    /// Convert the key material from `key`/`iv`, into a `ConnectionTrafficSecrets` item.
+    fn extract_keys(&self, key: &[u8], iv: &[u8], explicit: &[u8]) -> ConnectionTrafficSecrets;
+}
+
+/// How a TLS1.2 `key_block` is partitioned.
+///
+/// nb. ciphersuites with non-zero `mac_key_length` not currently supported
+pub struct KeyBlockShape {
+    /// How long keys are.
+    ///
+    /// `enc_key_len` terminology is from the standard.
```

Comment:
similar comment as earlier: replace "from the standard" with a more concrete pointer to rfc5246 A.6.